File: dxread/models/dxs3.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass
class DXS3:
    """The DXS3 data. 7 bytes
    daynit :  1;
    ithr   :  3;
    vthr   :  3;
    shore  :  1;
    iret   :  4;
    icsret :  4;
    icsrad :  8;
    itmp   :  8;
    iprs   :  8;
    icstmp :  8;
    icsprs :  8;
    """

    daynit: int
    ithr: int
    vthr: int
    shore: int
    iret: int
    icsret: int
    itmp: int
    iprs: int
    icstmp: int
    icsprs: int

    @staticmethod
    def from_bytes(br: bytes) -> DXS3:
        daynit = (br[0] & 0b10000000) >> 7
        ithr = (br[0] & 0b01110000) >> 4
        vthr = (br[0] & 0b00001110) >> 1
        shore = br[0] & 0b00000001
        iret = (br[1] & 0b11110000) >> 4
        icsret = br[1] & 0b00001111
        itmp = br[2]
        iprs = br[3]
        icstmp = br[4]
        icsprs = br[5]
        return DXS3(
            daynit=daynit,
            ithr=ithr,
            vthr=vthr,
            shore=shore,
            iret=iret,
            icsret=icsret,
            itmp=itmp,
            iprs=iprs,
            icstmp=icstmp,
            icsprs=icsprs,
        )

File: zig-dxread/src/models/dx_s3.zig
const std = @import("zig");
const constants = @import("../constants.zig");

pub const DXS3 = struct {
    daynit: u8,
    ithr: u8,
    vthr: u8,
    shore: u8,
    iret: u8,
    icsret: u8,
    itmp: u8,
    iprs: u8,
    icstmp: u8,
    icsprs: u8,

    pub fn fromBuffer(buffer: []const u8) DXS3 {
        return DXS3{
            .daynit = (buffer[0] & 0b10000000) >> 7,
            .ithr = (buffer[0] & 0b01110000) >> 4,
            .vthr = (buffer[0] & 0b00001110) >> 1,
            .shore = buffer[0] & 0b00000001,
            .iret = (buffer[1] & 0b11110000) >> 4,
            .icsret = buffer[1] & 0b00001111,
            .itmp = buffer[2],
            .iprs = buffer[3],
            .icstmp = buffer[4],
            .icsprs = buffer[5],
        };
    }

    pub fn deinit(self: *DXS3) void {
        self.* = undefined;
    }
};
